File: agora-easycall-ios-oc/AgoraSample/Menu/AGSMenuViewController.m
//
//  AGSMenuViewController.m
//  AgoraSample
//
//  Created by Shilong on 9/16/15.
//  Copyright (c) 2015 Agora. All rights reserved.
//

#import "AGSMenuViewController.h"
#import "AGSSelectMediaViewController.h"

@implementation AGSResolution

+ (instancetype)resolutionWithType:(AGSResolutionType)type
{
    AGSResolution *re = [[self alloc] init];
    if (type == AGSResolutionTypeCIF) {
        re.width = @352;
        re.height = @288;
    } else if (type == AGSResolutionTypeLow) {
        re.width = @640;
        re.height = @480;
    } else if (type == AGSResolutionTypeMiddle) {
        re.width = @1280;
        re.height = @720;
    } else {
        re.width = @1920;
        re.height = @1080;
    }
    return re;
}

+ (NSString *)resolutionFormatWithType:(AGSResolutionType)type;
{
    NSString *string = @"";
    if (type == AGSResolutionTypeCIF) {
        string = @"CIF";
    } else if (type == AGSResolutionTypeLow) {
        string = @"480P";
    } else if (type == AGSResolutionTypeMiddle) {
        string = @"720P";
    } else {
        string = @"1080P";
    }
    return string;
}

@end


@interface AGSMenuViewController ()

//
@property (weak, nonatomic) IBOutlet UILabel *menuTitleHistory;
@property (weak, nonatomic) IBOutlet UILabel *menuTitleResolution;
@property (weak, nonatomic) IBOutlet UILabel *menuTitleMaxRate;
@property (weak, nonatomic) IBOutlet UILabel *menuTitleMaxFrame;
@property (weak, nonatomic) IBOutlet UILabel *menuTitleVolume;
@property (weak, nonatomic) IBOutlet UILabel *menuTitleRecording;
@property (weak, nonatomic) IBOutlet UILabel *menuTitleLogPath;
@property (weak, nonatomic) IBOutlet UILabel *menuTitleFloatWindow;
@property (weak, nonatomic) IBOutlet UILabel *menuTitleAbout;

@property (weak, nonatomic) IBOutlet UILabel *nameLabel;
@property (weak, nonatomic) IBOutlet UILabel *idLabel;

@property (weak, nonatomic) IBOutlet UISlider *resolutionSlider;
@property (weak, nonatomic) IBOutlet UILabel *resolutionLabel;

@property (weak, nonatomic) IBOutlet UISlider *rateSlider;
@property (weak, nonatomic) IBOutlet UILabel *rateLabel;

@property (weak, nonatomic) IBOutlet UISlider *frameSlider;
@property (weak, nonatomic) IBOutlet UILabel *frameLabel;

@property (weak, nonatomic) IBOutlet UISwitch *maxRateSwitch;
@property (weak, nonatomic) IBOutlet UISwitch *recordingSwitch;
@property (weak, nonatomic) IBOutlet UISwitch *floatSwitch;

@property (weak, nonatomic) IBOutlet UISlider *volumeSlider;

@property (weak, nonatomic) IBOutlet UILabel *logFilePathLabel;

//@property (nonatomic) NSUserDefaults *defaults;

@property (nonatomic) AGSResolutionType currentResolutionType;
@property (nonatomic) int currentRate;
@property (nonatomic) int currentFrame;
@property (nonatomic) BOOL currentValueFloatWindow;
@property (nonatomic) BOOL currentValueRecording;

@end

@implementation AGSMenuViewController

#pragma mark -

- (IBAction)didValueChangedResolutionSlider:(UISlider *)sender {
    AGSResolutionType type = AGSResolutionTypeHigh;
    
    if (sender.value <= 0.25) {
        type = AGSResolutionTypeCIF;
    } else if (sender.value <= 0.5) {
        type = AGSResolutionTypeLow;
    } else if (sender.value <= 0.75) {
        type = AGSResolutionTypeMiddle;
    } else {
        type = AGSResolutionTypeHigh;
    }
    
    if (type != self.currentResolutionType) {
        self.currentResolutionType = type;
        self.resolutionLabel.text = [AGSResolution resolutionFormatWithType:type];
        if (self.resolutionBlock) {
            self.resolutionBlock([AGSResolution resolutionWithType:type]);
        }
    }
}

- (IBAction)didValueChangedRateSlider:(UISlider *)sender {
    float d = 1.0/7.0;
    int rate = 0;
    NSString *string = nil;
    if (sender.value <= d) {
        rate = 150 * 1024;
        string = @"150K";
    } else if (sender.value <= d * 2.0) {
        rate = 500 * 1024;
        string = @"500K";
    } else if (sender.value <= d * 3.0) {
        rate = 800 * 1024;
        string = @"800K";
    } else if (sender.value <= d * 4.0) {
        rate = 1024 * 1024;
        string = @"1M";
    } else if (sender.value <= d * 5.0) {
        rate = 2 * 1024 * 1024;
        string = @"2M";
    } else if (sender.value <= d * 6.0) {
        rate = 5 * 1024 * 1024;
        string = @"5M";
    } else {
        rate = 10 * 1024 *1024;
        string = @"10M";
    }
    
    if (rate != self.currentRate) {
        self.currentRate = rate;
        self.rateLabel.text = [NSString stringWithFormat:@"%@bps", string];
        if (self.rateBlock) {
            self.rateBlock(self.currentRate);
        }
    }
}

- (IBAction)didValueChangedFrameSlider:(UISlider *)sender {
    int frame = 0;
    if (sender.value <= 0.2) {
        frame = 15;
    } else if (sender.value <= 0.4) {
        frame = 20;
    } else if (sender.value <= 0.6) {
        frame = 24;
    } else if (sender.value <= 0.8) {
        frame = 30;
    } else {
        frame = 60;
    }
    
    if (frame != self.currentFrame) {
        self.currentFrame = frame;
        self.frameLabel.text = [NSString stringWithFormat:@"%dFPS", frame];
        if (self.frameBlock) {
            self.frameBlock(self.currentFrame);
        }
    }
}

- (IBAction)didValueChangedVolumeSlider:(UISlider *)sender {
    int volume = 255 * sender.value;
    
    if (self.volumeBlock) {
        self.volumeBlock(volume);
    }
    
}

- (IBAction)didValueChangedRecordingSwitch:(UISwitch *)sender {
    if (self.recordingBlock) {
        self.recordingBlock(sender.on);
    }
}

- (IBAction)didValueChangedFloatSwitch:(UISwitch *)sender {
    if (self.floatWindowBlock) {
        self.floatWindowBlock(sender.on);
    }
}

#pragma mark -

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.currentResolutionType = AGSResolutionTypeLow;
    self.resolutionLabel.text = @"480P";
    self.resolutionSlider.value = 0.3;
    
    self.currentFrame = 30;
    self.frameLabel.text = @"30FPS";
    self.frameSlider.value = 0.8;
    
    self.currentRate = 1024 * 1024;
    self.rateSlider.value = 4.0 * (1.0/7.0);
    self.rateLabel.text = @"1Mbps";
    
    self.volumeSlider.value = 100.0/255.0;
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    self.nameLabel.text = [userDefaults objectForKey:AGSKeyUsername];
    self.idLabel.text = [userDefaults objectForKey:AGSKeyVendorKey];
    
    //
    self.menuTitleAbout.text = NSLocalizedString(@"About", @"");
    self.menuTitleFloatWindow.text = NSLocalizedString(@"Pop up", @"");
    self.menuTitleHistory.text = NSLocalizedString(@"Conf. History", @"");
    
    self.menuTitleLogPath.text = NSLocalizedString(@"Log Path", @"");
    self.menuTitleMaxFrame.text = NSLocalizedString(@"Max. Framerate", @"");
    self.menuTitleMaxRate.text = NSLocalizedString(@"Max. Bitrate", @"");
    
    self.menuTitleRecording.text = NSLocalizedString(@"Recording", @"");
    self.menuTitleResolution.text = NSLocalizedString(@"Video Resolution", @"");
    self.menuTitleVolume.text = NSLocalizedString(@"Speaker Volume", @"");
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: agora-easycall-ios-swift/SampleSwift/AGSMenuViewController.swift
//
//  AGSMenuViewController.swift
//  SampleSwift
//
//  Created by let on 10/29/15.
//  Copyright © 2015 Agora. All rights reserved.
//

import UIKit

enum AGSResolutionType {
    case CIF, Low, Middle, High
    
    func description() -> String {
        switch self {
        case .CIF:
            return "CIF"
        case .Low:
            return "480P"
        case .Middle:
            return "720P"
        case .High:
            return "1080P"
        }
    }
    
    func width() -> Int32 {
        switch self {
        case .CIF:
            return 352
        case .Low:
            return 640
        case .Middle:
            return 1280
        case .High:
            return 1920
        }
    }
    
    func height() -> Int32 {
        switch self {
        case .CIF:
            return 288
        case .Low:
            return 480
        case .Middle:
            return 720
        case .High:
            return 1080
        }
    }
}

class AGSMenuViewController: UITableViewController {
    
    // Closures
    var resolutionChangedHandler: ((AGSResolutionType) -> Void)!
    var frameChangedHandler: ((Int32) -> Void)!
    var rateChangedHandler: ((Int32) -> Void)!
    var volumeChangedHandler: ((Int32) -> Void)!
    
    var recordingChangedHandler: ((Bool) -> Void)!
    var floatWindowChangedHandler: ((Bool) -> Void)!
    
    // IB Outlets
    
    //
    @IBOutlet var menuTitleHistory: UILabel!
    @IBOutlet var menuTitleResolution: UILabel!
    @IBOutlet var menuTitleMaxRate: UILabel!
    @IBOutlet var menuTitleMaxFrame: UILabel!
    @IBOutlet var menuTitleVolume: UILabel!
    @IBOutlet var menuTitleRecording: UILabel!
    @IBOutlet var menuTitleLogPath: UILabel!
    @IBOutlet var menuTitleFloatWindow: UILabel!
    @IBOutlet var menuTitleAbout: UILabel!
    
    //
    @IBOutlet var nameLabel: UILabel!
    @IBOutlet var idLabel: UILabel!
    @IBOutlet var resolutionSlider: UISlider!
    @IBOutlet var resolutionLabel: UILabel!
    @IBOutlet var rateSlider: UISlider!
    @IBOutlet var rateLabel: UILabel!
    @IBOutlet var frameSlider: UISlider!
    @IBOutlet var frameLabel: UILabel!
    @IBOutlet var recordingSwitch: UISwitch!
    @IBOutlet var floatSwitch: UISwitch!
    @IBOutlet var logFilePathLabel: UILabel!
    
    @IBOutlet var volumeSlider: UISlider!
    
    var currentResolutionType: AGSResolutionType?
    var currentRate: Int32?
    var currentFrame: Int32?
    var currentValueFloatWindow: Bool?
    var currentValueRecording: Bool?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        
        currentResolutionType = AGSResolutionType.Low
        resolutionLabel.text = currentResolutionType?.description()
        resolutionSlider.value = 0.3
        
        currentFrame = 30
        frameLabel.text = "30FPS"
        frameSlider.value = 0.8
        
        currentRate = 2014 * 1024
        rateLabel.text = "1Mbps"
        rateSlider.value = 4.0 * (1.0 / 7.0)
        
        volumeSlider.value = 100.0/255.0
    }

    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        let defaults = NSUserDefaults.standardUserDefaults()
        nameLabel.text = defaults.objectForKey(AGSKeys.username) as? String
        idLabel.text = defaults.objectForKey(AGSKeys.vendorKey) as? String
        
        //
        //
        menuTitleAbout.text = NSLocalizedString("About", comment:"");
        menuTitleFloatWindow.text = NSLocalizedString("Pop up", comment:"");
        menuTitleHistory.text = NSLocalizedString("Conf. History", comment:"");
        
        menuTitleLogPath.text = NSLocalizedString("Log Path", comment:"");
        menuTitleMaxFrame.text = NSLocalizedString("Max. Framerate", comment:"");
        menuTitleMaxRate.text = NSLocalizedString("Max. Bitrate", comment:"");
        
        menuTitleRecording.text = NSLocalizedString("Recording", comment:"");
        menuTitleResolution.text = NSLocalizedString("Video Resolution", comment:"");
        menuTitleVolume.text = NSLocalizedString("Speaker Volume", comment:"");
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: - IB Actions
    
    @IBAction func didValueChangedResolutionSlider(sender: UISlider) {
        var type: AGSResolutionType!
        
        if sender.value <= 0.25 {
            type = AGSResolutionType.CIF;
        } else if sender.value <= 0.5 {
            type = AGSResolutionType.Low;
        } else if sender.value <= 0.75 {
            type = AGSResolutionType.Middle;
        } else {
            type = AGSResolutionType.High;
        }
        
        if (type != currentResolutionType) {
            currentResolutionType = type
            resolutionLabel.text = type.description()
            if resolutionChangedHandler != nil {
                resolutionChangedHandler(type)
            }
        }
    }
    
    @IBAction func didValueChangedRateSlider(sender: UISlider) {
        let d: Float = 1.0/7.0
        var rate: Int32 = 0
        var string :String
        if sender.value <= d {
            rate = 150 * 1024
            string = "150K";
        } else if (sender.value <= d * 2.0) {
            rate = 500 * 1024;
            string = "500K";
        } else if (sender.value <= d * 3.0) {
            rate = 800 * 1024;
            string = "800K";
        } else if (sender.value <= d * 4.0) {
            rate = 1024 * 1024;
            string = "1M";
        } else if (sender.value <= d * 5.0) {
            rate = 2 * 1024 * 1024;
            string = "2M";
        } else if (sender.value <= d * 6.0) {
            rate = 5 * 1024 * 1024;
            string = "5M";
        } else {
            rate = 10 * 1024 * 1024;
            string = "10M";
        }
        
        if rate != currentRate {
            currentRate = rate
            rateLabel.text = string + "bps"
            if rateChangedHandler != nil {
                rateChangedHandler(currentRate!)
            }
        }
    }
    
    @IBAction func didValueChangedFrameSlider(sender: UISlider) {
        var frame: Int32 = 0
        if sender.value <= 0.2 {
            frame = 15;
        } else if sender.value <= 0.4 {
            frame = 20;
        } else if sender.value <= 0.6 {
            frame = 24;
        } else if sender.value <= 0.8 {
            frame = 30;
        } else {
            frame = 60;
        }
        
        if frame != currentFrame {
            currentFrame = frame
            frameLabel.text = frame.description + "FPS"
            if frameChangedHandler != nil {
                frameChangedHandler(currentFrame!)
            }
        }
    }

    @IBAction func didValueChangedVolumeSlider(sender: UISlider) {
        let volume = 255.0 * sender.value
        
        if volumeChangedHandler != nil {
            volumeChangedHandler(Int32(volume))
        }
    }
    
    @IBAction func didValueChangedRecordingSwitch(sender: UISwitch) {
        if recordingChangedHandler != nil {
            recordingChangedHandler(sender.on)
        }
    }
    
    @IBAction func didValueChangedFloatSwitch(sender: UISwitch) {
        if floatWindowChangedHandler != nil {
            floatWindowChangedHandler(sender.on)
        }
    }
    
//    func resolutionChanged(closure: (AGSResolutionType) -> Void) {
//        
//    }
//    
//    func frameChanged(closure: (Int) -> Void) {
//        
//    }
//    
//    func rateChanged(closure: (Int) -> Void) {
//        
//    }
//    
//    func recordingChanged(closure: (Bool) -> Void) {
//        
//    }
//    
//    func floatWindowChanged(closure: (Bool) -> Void) {
//        
//    }
    
    
    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
    }
    */

}
